docs/conf.py

```diff
@@ -233,6 +237,36 @@
 
 """
 
+
+# ----------------------------------------------------------------------------------
+# Redirects
+# ----------------------------------------------------------------------------------
+
+def determine_api_redirects() -> dict[str, str]:
+    """Set up API redirects for functions that we moved to module pages.
+
+    Note that we have redirects in Cloudflare for methods moving to their class page. We
+    could not do this for functions because some functions still have dedicated
+    HTML pages, so we cannot use a generic rule.
+    """
+    lines = Path("api_redirects.txt").read_text().splitlines()
+    result = {}
+    for line in lines:
+        if not line:
+            continue
+        old_doc_name, new_module_page_name = line.split(" ")
+        obj_name = old_doc_name.split("stubs/")[1]
+        # E.g. `../apidoc/assembler.html#qiskit.assembler.assemble_circuits
+        result[old_doc_name] = (
+            "https://qiskit.org/documentation/apidoc/" +
+            f"{new_module_page_name}.html#{obj_name}"
+        )
```

Comment:
> espite the pages mentioned in the flat file all being moved from stubs to apidoc, we have stubs in the file

Yeah, good point. I'll tweak.


> I'm also pretty nervous about hardcoding the deployment URL here - as an immediate effect, it means that local builds, and the dev and stable-branch deployments will all be broken, but as a general thing, Sphinx usually goes out of its way not to require this.

So my original implementation tried using a relative file, which is what sphinx-reredirects suggests using: https://documatt.gitlab.io/sphinx-reredirects. But the redirects did not actually work when I tested out the build. I suspect it can't handle the anchors properly.

I noticed we used full URLs in qiskit-metapackage for the Aer redirects: https://github.com/Qiskit/qiskit-metapackage/pull/1742/files, which led me to try it here.

Re: local builds, agreed. That's unfortunate that this won't handle local builds properly. I reason that it's worth the tradeoff so that we can have anchors, though.

Re: dev and stable, I don't think we care much about redirects on those? Sphinx already handles fixing bad links with internal docs via its cross references. What we're trying to fix with this PR is external links, like YouTube. If those were using a stable version, then there's nothing broken because we only changed the organization with a new Qiskit version and didn't change prior stable versions. We want to handle unstable URLs like https://qiskit.org/stubs/qiskit.algorithms.eval_observables.html.